### PR TITLE
Introduce an mc-transaction-extra crate taking bits of core and builder

### PR DESCRIPTION
**Pull request by garbageslam**
The thing that got me thinking about this is:
* When trying to build the reserve auditor, the build fails unless we set SGX_MODE and IAS_MODE, even though the reserve auditor does not do any attestation or check any IAS reports. That seems wierd. I investigated and figured out that `mc-api` depends on `mc-transaction-builder` to get the `UnsignedTx` object, because that has a counterpart in `external.proto`. I reported in #2630 that we should rejigger the dependency tree. (Actually, it's a separate question why transaction builder brings in the attest verifier code.)
* Initially my thought was that `UnsignedTx` should live in `mc-transaction-core` rather than `mc-transaction-builder`. However, that's actually not a great fix, because it forces the consensus-enclave-impl to depend on `UnsignedTx`, even though it doesn't use it.
* Still it seems reasonable that a project like reserve-auditor should be able to depend on `mc-api` without pulling in the `TransactionBuilder`, since it doesn't need to build any transactions. That feels like we are tightly coupling a huge amount of business logic to an API crate.

The more I thought about it, I became convinced that there needs to be a third crate in here, because there are a bunch of types that, are "important" and "relevant to transactions", and may even be specified in MCIPs and involve cryptography etc. But, they aren't directly needed for transaction validation, and they aren't part of defining what is a valid mobilecoin transaction.

My opinion is that `mc-transaction-core` should be limited to just the things that are required to be in the enclave for validation, to the extent possible.

* This lets us put as little code in the enclave as possible (good)
* This lets us easily see that changing a particular struct cannot break enclave compatibility (good)
* During PR review (e.g. [this one](https://github.com/mobilecoinfoundation/mobilecoin/pull/1871)) it becomes more clear that e.g. `InputRules` is part of the core and represents how "what is a valid transaction?" is changing, while `SignedContingentInput` is more like an "extra" schema, that isn't part of the core protocol, but rather some kind of layer 2 thing on top of the main protocol. (good)
* Possibly, this can help with splitting up changes like that, since you have to plan out what is core and what is extra, so you can do the core part and tests in one step and the extra parts in another.

I identified a number of things that can move from `mc-transaction-core` to this new `mc-transaction-extra` crate:

* `TxOutConfirmationNumber`
* `TxOutGiftCode`
* `SignedContingentInput`

I also identified some things in `mc-transaction-builder` that could be better off in the `mc-transaction-extra` crate:

* Memo types
* `UnsignedTx`

The memo builders remain in the builder crate.

This also fixes my original issue -- the `mc-api` crate no longer depends on `mc-transaction-builder`.

If we merge this, I would rebase MCIP 52 work, and make it so that:

* `TxSummary` (and friends) are part of `mc-transaction-core`
* `TxSummaryUnblindingData` (and friends) are part of `mc-transaction-extra`

This is just code movement, there are no functional changes.

### CODE REVIEW

**Comment by eranrund** [comment]
Thank you for taking the time to write a detailed explanation to the reasoning behind this. It seems very reasonable to me. 

> This also fixes my original issue -- the `mc-api` crate no longer depends on `mc-transaction-builder`.

Looking at `api/cargo.toml` I don't see the builder dependency removed...

**Comment by garbageslam** [comment]
thank you! should be fixed now

**Review by eranrund — APPROVED** [review]

**Review by remoun — APPROVED** [review]
LGTM

File: transaction/extra/src/tx_out_confirmation_number.rs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        tx_pubkey: &RistrettoPublic,
+        view_private_key: &RistrettoPrivate,
+    ) -> bool {
+        let shared_secret = get_tx_out_shared_secret(view_private_key, tx_pubkey);
+        let calculated_confirmation = TxOutConfirmationNumber::from(&shared_secret);
+        calculated_confirmation == *self
+    }
+}
+
+impl core::convert::AsRef<[u8; 32]> for TxOutConfirmationNumber {
```
Comment (jcape) [inline transaction/extra/src/tx_out_confirmation_number.rs:40]: Future reference: all these traits are part of `core::prelude`, so we shouldn't need to reference them absolutely.

**Review by jcape — APPROVED** [review]

File: transaction/extra/src/tx_out_confirmation_number.rs (hunk 1 above)
Comment (garbageslam) [reply transaction/extra/src/tx_out_confirmation_number.rs:40]: thanks, didn't know this

**Pull request merged**